core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,496 @@
+package io.grpc;
+
+import com.google.common.base.Preconditions;
+import com.google.common.util.concurrent.MoreExecutors;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+
+import javax.annotation.Nullable;
+
+
+/**
+ * --Add warning about using this too liberally--
+ * --Add warning about using this with containers that unload Classloaders--
+ * -- Add note that in the case of cascading closure that listeners to the parent are always
+ * notified before listeners on the children.
+ *
+ */
+public class Context {
+
+  /**
+   * Stack of context objects which is used to record attach & detach history on a thread.
+   */
+  private static final ThreadLocal<ArrayDeque<Context>> contextStack =
```

Comment:
The ContinuationLocal stuff is still a bit on the drawing board, for now it's just a rough port of Twitter's implementation, but that will probably change a little bit. We'll need to find a stable public API to build on. Our motivation for the ContinuationLocal standard is that it makes it much easier for implementors of tracing systems to rely on the same underlying constructs, given these constructs have wide adoption. In some ways it's similar to reactive streams. (Maybe it could also fit in the Java standard library later on)
The important part is that all libraries should use it. Say that you use rxjava - your grpc tracing will be broken because of the implicit thread switching that happens in rxjava. Or ditto for Scala futures. But (conceivably) it would work with a ContinuationLocal, because projects like rxjava and Scala decided to support ContinuationLocals natively. (Given it were the industry standard)

Here's some considerations about the list of differences:

1) Giving them names could be a handy opt-in feature we could add to ContinuationLocals. Or one can use ContinuationLocal<Map<String,Object>> if one wants to do it that way.

2) Do you mean something like this?

```java
myVar.setValue("foo");
System.out.println(myVar.get()); // foo
myVar.let("bar", () -> {
    System.out.println(myVar.get()); // bar
});
System.out.println(myVar.get()); // foo
```

where you can override the value within a given context. The save() and restore() methods in ContinuationLocal are meant to facilitate this kind of overlaying state.

3) ContinuationLocal comes with utilities for context passing (save() and .restore(), and the ExecutorService wrapper), the idea would be that there is one industry-standard way in Java to pass state between threads.

4) Cancellation is out of scope for the ContinuationLocal, but one could maybe use ContinationLocal instead of ThreadLocal as the underlying storage mechanism in Context, and then instead of providing a custom utility for cross-thread stuff, use the ContinuationLocal ones?